src/altair/block_processing.rs

```diff
@@ -55,9 +175,73 @@ pub fn process_deposit<
         MAX_VALIDATORS_PER_COMMITTEE,
         SYNC_COMMITTEE_SIZE,
     >,
-    _deposit: &mut Deposit,
-    _context: &Context,
+    deposit: &mut Deposit,
+    context: &Context,
 ) -> Result<()> {
+    let branch = deposit
+        .proof
+        .iter()
+        .map(|node| Node::from_bytes(node.as_ref().try_into().unwrap()))
+        .collect::<Vec<_>>();
+    let leaf = deposit.data.hash_tree_root()?;
+    let depth = DEPOSIT_CONTRACT_TREE_DEPTH + 1;
+    let index = state.eth1_deposit_index as usize;
+    let root = &state.eth1_data.deposit_root;
+    if !is_valid_merkle_branch(&leaf, branch.iter(), depth, index, root) {
+        return Err(invalid_operation_error(InvalidOperation::Deposit(
+            InvalidDeposit::InvalidProof {
+                leaf,
+                branch,
+                depth,
+                index,
+                root: *root,
+            },
+        )));
+    }
+
+    // NOTE: deviate from the order of the spec to avoid mutations
+    // that would need to be rolled back upon failure
+    let public_key = deposit.data.public_key.clone();
+    let amount = deposit.data.amount;
+    let validator_public_keys: HashSet<&BlsPublicKey> =
+        HashSet::from_iter(state.validators.iter().map(|v| &v.public_key));
+    if !validator_public_keys.contains(&public_key) {
+        let mut deposit_message = DepositMessage {
+            public_key: public_key.clone(),
+            withdrawal_credentials: deposit.data.withdrawal_credentials.clone(),
+            amount,
+        };
+        let domain = compute_domain(DomainType::Deposit, None, None, context)?;
+        let signing_root = compute_signing_root(&mut deposit_message, domain)?;
+        // Initialize validator if the deposit signature is valid
+        if public_key.verify_signature(signing_root.as_bytes(), &deposit.data.signature) {
+            state
+                .validators
+                .push(get_validator_from_deposit(deposit, context));
+            state.balances.push(amount);
+            state
+                .previous_epoch_participation
+                .push(ParticipationFlags::default());
+            state
+                .current_epoch_participation
+                .push(ParticipationFlags::default());
+            state.inactivity_scores.push(u64::default())
```

Comment:
I have a weak preference for this 
```suggestion
            state.inactivity_scores.push(0)
```

but tbh the compiler probably compiles both to the same code